File: JsCode/leetcode/easy/LetterCasePermutation.js
/**
 * Leetcode 784. Letter Case Permutation
 * https://leetcode.com/problems/letter-case-permutation/description/
 * 数组的方法splice方法：
 * arrayObject.splice(index,howmany,item1,.....,itemX)
 * splice() 方法可删除从 index 处开始的零个或多个元素，并且用参数列表中声明的一个或多个值来替换那些被删除的元素。
 * 如果从 arrayObject 中删除了元素，则返回的是含有被删除的元素的数组
 * @param {string} S
 * @return {string[]}
 */
var letterCasePermutation = function (S) {
    var lowS = S.toLowerCase();
    var resultList = [''];
    for (var i = 0; i < lowS.length; i++) {
        if (lowS[i] >= '0' && lowS[i] <= '9') {
            for (var k = 0; k < resultList.length; k++) {
                resultList[k] += lowS[i];
            }
        } else {
            var size = resultList.length;
            for (var j = 0; j < 2 * size; j += 2) {
                var temp = resultList[j];
                resultList[j] += lowS[i];
                resultList.splice(j + 1, 0, temp + lowS[i].toUpperCase());
            }
        }
    }
    return resultList;
};
var S = 'a1b2';
var resultList = letterCasePermutation(S);
for (var i = 0; i < resultList.length; i++) {
    console.log(resultList[i]);
}

File: PythonCode/leetcode/easy/LetterCasePermutation.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

# @Time    : 2018/2/28 11:25
# @Author  : bearyang
# @Site    : 
# @File    : LetterCasePermutation.py
# @Software: PyCharm
# Leetcode 784. Letter Case Permutation
# https://leetcode.com/problems/letter-case-permutation/description/


class Solution:
    def letterCasePermutation(self, S):
        """
        :type S: str
        :rtype: List[str]
        """
        lowS = S.lower()
        resultList = ['']
        for ch in lowS:
            if ch.isdigit():
                for i in range(0, len(resultList)):
                    resultList[i] += ch
            else:
                size = len(resultList)
                for j in range(0, 2 * size, 2):
                    temp = resultList[j]
                    resultList[j] += ch
                    resultList.insert(j + 1, temp + ch.upper())
        return resultList


test = Solution()
S = 'a1b2'
resultList = test.letterCasePermutation(S)
for item in resultList:
    print(item)